File: app/src/main/java/ru/melod1n/vk/database/CacheStorage.java
package ru.melod1n.vk.database;

import android.content.ContentValues;
import android.database.Cursor;

import java.util.ArrayList;

import ru.melod1n.vk.api.UserConfig;
import ru.melod1n.vk.api.model.VKConversation;
import ru.melod1n.vk.api.model.VKGroup;
import ru.melod1n.vk.api.model.VKMessage;
import ru.melod1n.vk.api.model.VKModel;
import ru.melod1n.vk.api.model.VKUser;
import ru.melod1n.vk.util.Util;

import static ru.melod1n.vk.common.AppGlobal.database;
import static ru.melod1n.vk.database.DatabaseHelper.ACTION;
import static ru.melod1n.vk.database.DatabaseHelper.ATTACHMENTS;
import static ru.melod1n.vk.database.DatabaseHelper.CAN_ACCESS_CLOSED;
import static ru.melod1n.vk.database.DatabaseHelper.CAN_WRITE;
import static ru.melod1n.vk.database.DatabaseHelper.CHAT_SETTINGS;
import static ru.melod1n.vk.database.DatabaseHelper.CLOSED;
import static ru.melod1n.vk.database.DatabaseHelper.CONVERSATION_ID;
import static ru.melod1n.vk.database.DatabaseHelper.DATE;
import static ru.melod1n.vk.database.DatabaseHelper.DEACTIVATED;
import static ru.melod1n.vk.database.DatabaseHelper.FIRST_NAME;
import static ru.melod1n.vk.database.DatabaseHelper.FRIEND_ID;
import static ru.melod1n.vk.database.DatabaseHelper.FROM_ID;
import static ru.melod1n.vk.database.DatabaseHelper.FWD_MESSAGES;
import static ru.melod1n.vk.database.DatabaseHelper.GROUP_ID;
import static ru.melod1n.vk.database.DatabaseHelper.IMPORTANT;
import static ru.melod1n.vk.database.DatabaseHelper.IN_READ;
import static ru.melod1n.vk.database.DatabaseHelper.IS_CLOSED;
import static ru.melod1n.vk.database.DatabaseHelper.LAST_NAME;
import static ru.melod1n.vk.database.DatabaseHelper.LAST_SEEN;
import static ru.melod1n.vk.database.DatabaseHelper.MESSAGE_ID;
import static ru.melod1n.vk.database.DatabaseHelper.NAME;
import static ru.melod1n.vk.database.DatabaseHelper.ONLINE;
import static ru.melod1n.vk.database.DatabaseHelper.ONLINE_MOBILE;
import static ru.melod1n.vk.database.DatabaseHelper.OUT_READ;
import static ru.melod1n.vk.database.DatabaseHelper.PEER;
import static ru.melod1n.vk.database.DatabaseHelper.PEER_ID;
import static ru.melod1n.vk.database.DatabaseHelper.PHOTO_100;
import static ru.melod1n.vk.database.DatabaseHelper.PHOTO_200;
import static ru.melod1n.vk.database.DatabaseHelper.PHOTO_50;
import static ru.melod1n.vk.database.DatabaseHelper.PUSH_SETTINGS;
import static ru.melod1n.vk.database.DatabaseHelper.RANDOM_ID;
import static ru.melod1n.vk.database.DatabaseHelper.REPLY_MESSAGE;
import static ru.melod1n.vk.database.DatabaseHelper.SCREEN_NAME;
import static ru.melod1n.vk.database.DatabaseHelper.SEX;
import static ru.melod1n.vk.database.DatabaseHelper.STATUS;
import static ru.melod1n.vk.database.DatabaseHelper.TABLE_CONVERSATIONS;
import static ru.melod1n.vk.database.DatabaseHelper.TABLE_FRIENDS;
import static ru.melod1n.vk.database.DatabaseHelper.TABLE_GROUPS;
import static ru.melod1n.vk.database.DatabaseHelper.TABLE_MESSAGES;
import static ru.melod1n.vk.database.DatabaseHelper.TABLE_USERS;
import static ru.melod1n.vk.database.DatabaseHelper.TEXT;
import static ru.melod1n.vk.database.DatabaseHelper.TYPE;
import static ru.melod1n.vk.database.DatabaseHelper.UNREAD_COUNT;
import static ru.melod1n.vk.database.DatabaseHelper.USER_ID;
import static ru.melod1n.vk.database.DatabaseHelper.VERIFIED;

public class CacheStorage {

    private static Cursor selectCursor(String table, String column, Object value) {
        return QueryBuilder.query()
                .select("*").from(table)
                .where(column.concat(" = ").concat(String.valueOf(value)))
                .asCursor(database);
    }

    private static Cursor selectCursor(String table, String where) {
        return QueryBuilder.query()
                .select("*").from(table).where(where)
                .asCursor(database);
    }

    private static Cursor selectCursor(String table) {
        return QueryBuilder.query()
                .select("*").from(table)
                .asCursor(database);
    }

    private static int getInt(Cursor cursor, String columnName) {
        return cursor.getInt(cursor.getColumnIndex(columnName));
    }

    private static String getString(Cursor cursor, String columnName) {
        return cursor.getString(cursor.getColumnIndex(columnName));
    }

    private static byte[] getBlob(Cursor cursor, String columnName) {
        return cursor.getBlob(cursor.getColumnIndex(columnName));
    }

    private static void insert(String table, ArrayList values) {
        database.beginTransaction();

        ContentValues cv = new ContentValues();
        for (int i = 0; i < values.size(); i++) {
            VKModel item = (VKModel) values.get(i);

            switch (table) {
                case TABLE_MESSAGES:
                    putValues((VKMessage) item, cv);
                    break;
                case TABLE_CONVERSATIONS:
                    putValues((VKConversation) item, cv);
                    break;
                case TABLE_USERS:
                    putValues((VKUser) item, cv, false);
                    break;
                case TABLE_FRIENDS:
                    putValues((VKUser) item, cv, true);
                    break;
                case TABLE_GROUPS:
                    putValues((VKGroup) item, cv);
                    break;
            }

            database.insert(table, null, cv);
            cv.clear();
        }

        database.setTransactionSuccessful();
        database.endTransaction();
    }

    public static void delete(String table, Object where, Object args) {
        database.delete(table, where + " = ?", new String[]{String.valueOf(args)});
    }

    public static void delete(String table) {
        database.delete(table, null, null);
    }

    public static ArrayList<VKMessage> getMessages() {
        Cursor cursor = selectCursor(TABLE_MESSAGES);

        ArrayList<VKMessage> messages = new ArrayList<>(cursor.getCount());

        while (cursor.moveToNext()) {
            messages.add(parseMessage(cursor));
        }

        cursor.close();

        return messages;
    }

    public static VKMessage getMessageByPeerId(int id) {
        Cursor cursor = selectCursor(TABLE_MESSAGES, PEER_ID, id);

        if (cursor.getCount() == 0) return null;

        cursor.moveToFirst();

        VKMessage message = parseMessage(cursor);

        cursor.close();

        return message;
    }

    public static ArrayList<VKConversation> getConversations() {
        return getConversations(0);
    }

    public static ArrayList<VKConversation> getConversations(int size) {
        Cursor cursor = selectCursor(TABLE_CONVERSATIONS);

        ArrayList<VKConversation> conversations = new ArrayList<>();

        while (cursor.moveToNext()) {
            conversations.add(parseConversation(cursor));
            if (conversations.size() == size && size > 0) break;
        }

        cursor.close();

        return conversations;
    }

    public static VKUser getUser(int id) {
        Cursor cursor = selectCursor(TABLE_USERS, USER_ID, id);

        if (cursor.getCount() == 0) return null;

        cursor.moveToFirst();

        VKUser user = parseUser(cursor);

        cursor.close();

        return user;
    }

    public static VKGroup getGroup(int id) {
        Cursor cursor = selectCursor(TABLE_GROUPS, GROUP_ID, id);

        if (cursor.getCount() == 0) return null;

        cursor.moveToFirst();

        VKGroup group = parseGroup(cursor);

        cursor.close();

        return group;
    }

    private static VKMessage parseMessage(Cursor cursor) {
        VKMessage message = new VKMessage();

        message.setId(getInt(cursor, MESSAGE_ID));
        message.setDate(getInt(cursor, DATE));
        message.setPeerId(getInt(cursor, PEER_ID));
        message.setFromId(getInt(cursor, FROM_ID));
        message.setText(getString(cursor, TEXT));
        message.setRandomId(getInt(cursor, RANDOM_ID));

        Object attachments = Util.deserialize(getBlob(cursor, ATTACHMENTS));
        message.setAttachments((ArrayList<VKModel>) attachments);

        message.setImportant(getInt(cursor, IMPORTANT) == 1);

        Object fwdMessages = Util.deserialize(getBlob(cursor, FWD_MESSAGES));
        message.setFwdMessages((ArrayList<VKMessage>) fwdMessages);

        Object replyMessage = Util.deserialize(getBlob(cursor, REPLY_MESSAGE));
        message.setReplyMessage((VKMessage) replyMessage);

        Object action = Util.deserialize(getBlob(cursor, ACTION));
        message.setAction((VKMessage.Action) action);

        return message;
    }

    private static VKConversation parseConversation(Cursor cursor) {
        VKConversation conversation = new VKConversation();

        Object peer = Util.deserialize(getBlob(cursor, PEER));
        conversation.setPeer((VKConversation.Peer) peer);

        conversation.setInRead(getInt(cursor, IN_READ));
        conversation.setOutRead(getInt(cursor, OUT_READ));
        conversation.setUnreadCount(getInt(cursor, UNREAD_COUNT));

        Object pushSettings = Util.deserialize(getBlob(cursor, PUSH_SETTINGS));
        conversation.setPushSettings((VKConversation.PushSettings) pushSettings);

        Object canWrite = Util.deserialize(getBlob(cursor, CAN_WRITE));
        conversation.setCanWrite((VKConversation.CanWrite) canWrite);

        Object chatSettings = Util.deserialize(getBlob(cursor, CHAT_SETTINGS));
        conversation.setChatSettings((VKConversation.ChatSettings) chatSettings);

        return conversation;
    }

    private static VKUser parseUser(Cursor cursor) {
        VKUser user = new VKUser();

        user.setId(getInt(cursor, USER_ID));
        user.setFirstName(getString(cursor, FIRST_NAME));
        user.setLastName(getString(cursor, LAST_NAME));
        user.setDeactivated(getString(cursor, DEACTIVATED));
        user.setClosed(getInt(cursor, CLOSED) == 1);
        user.setCanAccessClosed(getInt(cursor, CAN_ACCESS_CLOSED) == 1);
        user.setSex(getInt(cursor, SEX));
        user.setScreenName(getString(cursor, SCREEN_NAME));
        user.setPhoto50(getString(cursor, PHOTO_50));
        user.setPhoto100(getString(cursor, PHOTO_100));
        user.setPhoto200(getString(cursor, PHOTO_200));
        user.setOnline(getInt(cursor, ONLINE) == 1);
        user.setOnlineMobile(getInt(cursor, ONLINE_MOBILE) == 1);
        user.setStatus(getString(cursor, STATUS));

        Object lastSeen = Util.deserialize(getBlob(cursor, LAST_SEEN));
        user.setLastSeen((VKUser.LastSeen) lastSeen);

        user.setVerified(getInt(cursor, VERIFIED) == 1);

        return user;
    }

    private static VKGroup parseGroup(Cursor cursor) {
        VKGroup group = new VKGroup();

        group.setId(getInt(cursor, GROUP_ID));
        group.setName(getString(cursor, NAME));
        group.setScreenName(getString(cursor, SCREEN_NAME));
        group.setIsClosed(getInt(cursor, IS_CLOSED));
        group.setDeactivated(getString(cursor, DEACTIVATED));
        group.setPhoto50(getString(cursor, PHOTO_50));
        group.setPhoto100(getString(cursor, PHOTO_100));
        group.setPhoto200(getString(cursor, PHOTO_200));

        return group;
    }

    private static void putValues(VKMessage message, ContentValues values) {
        values.put(MESSAGE_ID, message.getId());
        values.put(DATE, message.getDate());
        values.put(PEER_ID, message.getPeerId());
        values.put(FROM_ID, message.getFromId());
        values.put(TEXT, message.getText());
        values.put(RANDOM_ID, message.getRandomId());

        if (message.getAttachments() != null) {
            values.put(ATTACHMENTS, Util.serialize(message.getAttachments()));
        }

        values.put(IMPORTANT, message.isImportant());

        if (message.getFwdMessages() != null) {
            values.put(FWD_MESSAGES, Util.serialize(message.getFwdMessages()));
        }

        if (message.getReplyMessage() != null) {
            values.put(REPLY_MESSAGE, Util.serialize(message.getReplyMessage()));
        }

        if (message.getAction() != null) {
            values.put(ACTION, Util.serialize(message.getAction()));
        }
    }

    private static void putValues(VKConversation conversation, ContentValues values) {
        if (conversation.getPeer() != null) {
            values.put(CONVERSATION_ID, conversation.getPeer().getId());
            values.put(PEER, Util.serialize(conversation.getPeer()));
        }

        values.put(IN_READ, conversation.getInRead());
        values.put(OUT_READ, conversation.getOutRead());
        values.put(UNREAD_COUNT, conversation.getUnreadCount());

        if (conversation.getPushSettings() != null) {
            values.put(PUSH_SETTINGS, Util.serialize(conversation.getPushSettings()));
        }

        if (conversation.getCanWrite() != null) {
            values.put(CAN_WRITE, Util.serialize(conversation.getCanWrite()));
        }

        if (conversation.getChatSettings() != null) {
            values.put(CHAT_SETTINGS, Util.serialize(conversation.getChatSettings()));
        }
    }

    private static void putValues(VKUser user, ContentValues values, boolean isFriend) {
        if (isFriend) {
            values.put(USER_ID, UserConfig.getUserId());
            values.put(FRIEND_ID, user.getId());
            return;
        }

        values.put(USER_ID, user.getId());
        values.put(FIRST_NAME, user.getFirstName());
        values.put(LAST_NAME, user.getLastName());
        values.put(DEACTIVATED, user.getDeactivated());
        values.put(CLOSED, user.isClosed());
        values.put(CAN_ACCESS_CLOSED, user.isCanAccessClosed());
        values.put(SEX, user.getSex());
        values.put(SCREEN_NAME, user.getScreenName());
        values.put(PHOTO_50, user.getPhoto50());
        values.put(PHOTO_100, user.getPhoto100());
        values.put(PHOTO_200, user.getPhoto200());
        values.put(ONLINE, user.isOnline());
        values.put(ONLINE_MOBILE, user.isOnlineMobile());
        values.put(STATUS, user.getStatus());

        if (user.getLastSeen() != null) {
            values.put(LAST_SEEN, Util.serialize(user.getLastSeen()));
        }

        values.put(VERIFIED, user.isVerified());
    }

    private static void putValues(VKGroup group, ContentValues values) {
        values.put(GROUP_ID, Math.abs(group.getId()));
        values.put(NAME, group.getName());
        values.put(SCREEN_NAME, group.getScreenName());
        values.put(IS_CLOSED, group.getIsClosed());
        values.put(DEACTIVATED, group.getDeactivated());
        values.put(TYPE, group.getType());
        values.put(PHOTO_50, group.getPhoto50());
        values.put(PHOTO_100, group.getPhoto100());
        values.put(PHOTO_200, group.getPhoto200());
    }

    public static void insertMessages(ArrayList<VKMessage> messages) {
        insert(TABLE_MESSAGES, messages);
    }

    public static void insertConversations(ArrayList<VKConversation> conversations) {
        insert(TABLE_CONVERSATIONS, conversations);
    }

    public static void insertUsers(ArrayList<VKUser> users) {
        insert(TABLE_USERS, users);
    }

    public static void insertGroups(ArrayList<VKGroup> groups) {
        insert(TABLE_GROUPS, groups);
    }
}

File: app/src/main/java/ru/melod1n/vk/database/CacheStorage.kt
package ru.melod1n.vk.database

import android.content.ContentValues
import android.database.Cursor
import android.text.TextUtils
import ru.melod1n.vk.api.UserConfig
import ru.melod1n.vk.api.model.*
import ru.melod1n.vk.common.AppGlobal
import ru.melod1n.vk.database.DatabaseHelper.Companion.ACTION
import ru.melod1n.vk.database.DatabaseHelper.Companion.ATTACHMENTS
import ru.melod1n.vk.database.DatabaseHelper.Companion.CAN_ACCESS_CLOSED
import ru.melod1n.vk.database.DatabaseHelper.Companion.CLOSED
import ru.melod1n.vk.database.DatabaseHelper.Companion.CONVERSATION_MESSAGE_ID
import ru.melod1n.vk.database.DatabaseHelper.Companion.DATE
import ru.melod1n.vk.database.DatabaseHelper.Companion.DEACTIVATED
import ru.melod1n.vk.database.DatabaseHelper.Companion.DISABLED_UNTIL
import ru.melod1n.vk.database.DatabaseHelper.Companion.EDIT_TIME
import ru.melod1n.vk.database.DatabaseHelper.Companion.FIRST_NAME
import ru.melod1n.vk.database.DatabaseHelper.Companion.FRIEND_ID
import ru.melod1n.vk.database.DatabaseHelper.Companion.FROM_ID
import ru.melod1n.vk.database.DatabaseHelper.Companion.FWD_MESSAGES
import ru.melod1n.vk.database.DatabaseHelper.Companion.GROUP_ID
import ru.melod1n.vk.database.DatabaseHelper.Companion.IMPORTANT
import ru.melod1n.vk.database.DatabaseHelper.Companion.IN_READ
import ru.melod1n.vk.database.DatabaseHelper.Companion.IS_ALLOWED
import ru.melod1n.vk.database.DatabaseHelper.Companion.IS_CLOSED
import ru.melod1n.vk.database.DatabaseHelper.Companion.IS_DISABLED_FOREVER
import ru.melod1n.vk.database.DatabaseHelper.Companion.IS_GROUP_CHANNEL
import ru.melod1n.vk.database.DatabaseHelper.Companion.IS_NO_SOUND
import ru.melod1n.vk.database.DatabaseHelper.Companion.LAST_MESSAGE_ID
import ru.melod1n.vk.database.DatabaseHelper.Companion.LAST_NAME
import ru.melod1n.vk.database.DatabaseHelper.Companion.LAST_SEEN
import ru.melod1n.vk.database.DatabaseHelper.Companion.LOCAL_ID
import ru.melod1n.vk.database.DatabaseHelper.Companion.MEMBERS_COUNT
import ru.melod1n.vk.database.DatabaseHelper.Companion.MESSAGE_ID
import ru.melod1n.vk.database.DatabaseHelper.Companion.NAME
import ru.melod1n.vk.database.DatabaseHelper.Companion.ONLINE
import ru.melod1n.vk.database.DatabaseHelper.Companion.ONLINE_MOBILE
import ru.melod1n.vk.database.DatabaseHelper.Companion.OUT
import ru.melod1n.vk.database.DatabaseHelper.Companion.OUT_READ
import ru.melod1n.vk.database.DatabaseHelper.Companion.PEER_ID
import ru.melod1n.vk.database.DatabaseHelper.Companion.PHOTO_100
import ru.melod1n.vk.database.DatabaseHelper.Companion.PHOTO_200
import ru.melod1n.vk.database.DatabaseHelper.Companion.PHOTO_50
import ru.melod1n.vk.database.DatabaseHelper.Companion.PINNED_MESSAGE_ID
import ru.melod1n.vk.database.DatabaseHelper.Companion.PLATFORM
import ru.melod1n.vk.database.DatabaseHelper.Companion.RANDOM_ID
import ru.melod1n.vk.database.DatabaseHelper.Companion.READ
import ru.melod1n.vk.database.DatabaseHelper.Companion.REASON
import ru.melod1n.vk.database.DatabaseHelper.Companion.REPLY_MESSAGE_ID
import ru.melod1n.vk.database.DatabaseHelper.Companion.SCREEN_NAME
import ru.melod1n.vk.database.DatabaseHelper.Companion.SEX
import ru.melod1n.vk.database.DatabaseHelper.Companion.STATE
import ru.melod1n.vk.database.DatabaseHelper.Companion.STATUS
import ru.melod1n.vk.database.DatabaseHelper.Companion.TABLE_CONVERSATIONS
import ru.melod1n.vk.database.DatabaseHelper.Companion.TABLE_FRIENDS
import ru.melod1n.vk.database.DatabaseHelper.Companion.TABLE_GROUPS
import ru.melod1n.vk.database.DatabaseHelper.Companion.TABLE_MESSAGES
import ru.melod1n.vk.database.DatabaseHelper.Companion.TABLE_USERS
import ru.melod1n.vk.database.DatabaseHelper.Companion.TEXT
import ru.melod1n.vk.database.DatabaseHelper.Companion.TITLE
import ru.melod1n.vk.database.DatabaseHelper.Companion.TYPE
import ru.melod1n.vk.database.DatabaseHelper.Companion.UNREAD_COUNT
import ru.melod1n.vk.database.DatabaseHelper.Companion.USER_ID
import ru.melod1n.vk.database.DatabaseHelper.Companion.VERIFIED
import ru.melod1n.vk.util.Util
import kotlin.math.abs


@Suppress("UNCHECKED_CAST")
object CacheStorage {
    private fun selectCursor(table: String, column: String, value: Any): Cursor {
        return QueryBuilder.query()
                .select("*").from(table)
                .where("$column = $value")
                .asCursor(AppGlobal.database)
    }

    private fun selectCursor(table: String, where: String): Cursor {
        return QueryBuilder.query()
                .select("*").from(table).where(where)
                .asCursor(AppGlobal.database)
    }

    private fun selectCursor(table: String): Cursor {
        return QueryBuilder.query()
                .select("*").from(table)
                .asCursor(AppGlobal.database)
    }

    private fun getInt(cursor: Cursor, columnName: String): Int {
        return cursor.getInt(cursor.getColumnIndex(columnName))
    }

    private fun getString(cursor: Cursor, columnName: String): String? {
        return cursor.getString(cursor.getColumnIndex(columnName))
    }

    private fun getBlob(cursor: Cursor, columnName: String): ByteArray? {
        return cursor.getBlob(cursor.getColumnIndex(columnName))
    }

    private fun insert(table: String, values: ArrayList<*>) {
        AppGlobal.database.beginTransaction()

        for (i in values.indices) {
            val cv = ContentValues()

            val item = values[i] as VKModel

            when (table) {
                TABLE_MESSAGES -> putValues(item as VKMessage, cv)
                TABLE_CONVERSATIONS -> putValues(item as VKConversation, cv)
                TABLE_USERS -> putValues(item as VKUser, cv, false)
                TABLE_FRIENDS -> putValues(item as VKUser, cv, true)
                TABLE_GROUPS -> putValues(item as VKGroup, cv)
            }

            AppGlobal.database.insert(table, null, cv)
            cv.clear()
        }

        AppGlobal.database.apply {
            setTransactionSuccessful()
            endTransaction()
        }
    }

    private fun update(table: String, values: ArrayList<*>, whereClause: String, whereArgs: Array<String>) {
        AppGlobal.database.beginTransaction()

        for (i in values.indices) {
            val cv = ContentValues()

            val item = values[i] as VKModel

            when (table) {
                TABLE_MESSAGES -> putValues(item as VKMessage, cv)
                TABLE_CONVERSATIONS -> putValues(item as VKConversation, cv)
                TABLE_USERS -> putValues(item as VKUser, cv, false)
                TABLE_FRIENDS -> putValues(item as VKUser, cv, true)
                TABLE_GROUPS -> putValues(item as VKGroup, cv)
            }

            AppGlobal.database.update(table, cv, "$whereClause = ?", whereArgs)
            cv.clear()
        }

        AppGlobal.database.apply {
            setTransactionSuccessful()
            endTransaction()
        }
    }

    fun delete(table: String?, where: Any, args: Any) {
        AppGlobal.database.delete(table, "$where = ?", arrayOf(args.toString()))
    }

    fun delete(table: String?) {
        AppGlobal.database.delete(table, null, null)
    }

    fun getMessage(messageId: Int): VKMessage? {
        val cursor = selectCursor(TABLE_MESSAGES, MESSAGE_ID, messageId)

        if (cursor.count == 0) return null

        cursor.moveToFirst()

        val message = parseMessage(cursor)

        cursor.close()

        return message
    }

    fun getMessages(peerId: Int): ArrayList<VKMessage> {
        val cursor = selectCursor(TABLE_MESSAGES, PEER_ID, peerId)

        val messages = ArrayList<VKMessage>(cursor.count)

        while (cursor.moveToNext()) {
            messages.add(parseMessage(cursor))
        }

        cursor.close()
        return messages
    }

    fun getConversation(peerId: Int): VKConversation? {
        val cursor = selectCursor(TABLE_CONVERSATIONS, PEER_ID, peerId)

        if (cursor.count == 0) return null

        cursor.moveToFirst()

        val conversation = parseConversation(cursor)

        cursor.close()

        return conversation
    }

    fun getConversations(): ArrayList<VKConversation> {
        return getConversations(0)
    }

    fun getConversations(size: Int): ArrayList<VKConversation> {
        val cursor = selectCursor(TABLE_CONVERSATIONS)
        val conversations = ArrayList<VKConversation>()

        while (cursor.moveToNext()) {
            conversations.add(parseConversation(cursor))
            if (conversations.size == size && size > 0) break
        }

        cursor.close()
        return conversations
    }

    fun getFriends(userId: Int, onlyOnline: Boolean): ArrayList<VKUser> {
        val cursor = QueryBuilder.query()
                .select("*")
                .from(TABLE_FRIENDS)
                .leftJoin(TABLE_USERS)
                .on("friends.friend_id = users.user_id")
                .where("friends.user_id = $userId")
                .asCursor(AppGlobal.database)

        val users: ArrayList<VKUser> = ArrayList(cursor.count)

        while (cursor.moveToNext()) {
            val userOnline = getInt(cursor, ONLINE) == 1

            if (onlyOnline && !userOnline) {
                continue
            }

            val user = parseUser(cursor)
            users.add(user)
        }

        cursor.close()
        return users
    }

    fun getUser(id: Int): VKUser? {
        val cursor = selectCursor(TABLE_USERS, USER_ID, id)
        if (cursor.count == 0) return null

        cursor.moveToFirst()

        val user = parseUser(cursor)

        cursor.close()
        return user
    }

    fun getGroup(gid: Int): VKGroup? {
        val id = abs(gid)
        val cursor = selectCursor(TABLE_GROUPS, GROUP_ID, id)
        if (cursor.count == 0) return null

        cursor.moveToFirst()

        val group = parseGroup(cursor)

        cursor.close()
        return group
    }

    private fun parseMessage(cursor: Cursor): VKMessage {
        return VKMessage().apply {
            id = getInt(cursor, MESSAGE_ID)
            date = getInt(cursor, DATE)
            peerId = getInt(cursor, PEER_ID)
            fromId = getInt(cursor, FROM_ID)
            editTime = getInt(cursor, EDIT_TIME)
            isOut = getInt(cursor, OUT) == 1
            conversationMessageId = getInt(cursor, CONVERSATION_MESSAGE_ID)
            text = getString(cursor, TEXT)
            randomId = getInt(cursor, RANDOM_ID)
            isRead = getInt(cursor, READ) == 1
            isImportant = getInt(cursor, IMPORTANT) == 1
            replyMessageId = getInt(cursor, REPLY_MESSAGE_ID)

            val attachments = Util.deserialize(getBlob(cursor, ATTACHMENTS))
            val fwdMessages = Util.deserialize(getBlob(cursor, FWD_MESSAGES))
            val action = Util.deserialize(getBlob(cursor, ACTION))

            this.fwdMessages = fwdMessages as ArrayList<VKMessage>?
            this.action = action as VKMessage.Action?
            this.attachments = attachments as ArrayList<VKModel>
        }
    }

    private fun parseConversation(cursor: Cursor): VKConversation {
        return VKConversation().apply {
            inRead = getInt(cursor, IN_READ)
            outRead = getInt(cursor, OUT_READ)
            unreadCount = getInt(cursor, UNREAD_COUNT)
            id = getInt(cursor, PEER_ID)
            isAllowed = getInt(cursor, IS_ALLOWED) == 1
            reason = getInt(cursor, REASON)
            lastMessageId = getInt(cursor, LAST_MESSAGE_ID)
            type = getString(cursor, TYPE)
            localId = getInt(cursor, LOCAL_ID)
            disabledUntil = getInt(cursor, DISABLED_UNTIL)
            isDisabledForever = getInt(cursor, IS_DISABLED_FOREVER) == 1
            isNoSound = getInt(cursor, IS_NO_SOUND) == 1
            membersCount = getInt(cursor, MEMBERS_COUNT)
            title = getString(cursor, TITLE)
            state = getString(cursor, STATE)
            isGroupChannel = getInt(cursor, IS_GROUP_CHANNEL) == 1
            photo50 = getString(cursor, PHOTO_50)
            photo100 = getString(cursor, PHOTO_100)
            photo200 = getString(cursor, PHOTO_200)

            pinnedMessageId = getInt(cursor, PINNED_MESSAGE_ID)
        }
    }

    private fun parseUser(cursor: Cursor): VKUser {
        return VKUser().apply {
            id = getInt(cursor, USER_ID)
            firstName = getString(cursor, FIRST_NAME)
            lastName = getString(cursor, LAST_NAME)
            deactivated = getString(cursor, DEACTIVATED)
            isClosed = getInt(cursor, CLOSED) == 1
            isCanAccessClosed = getInt(cursor, CAN_ACCESS_CLOSED) == 1
            sex = getInt(cursor, SEX)
            screenName = getString(cursor, SCREEN_NAME)
            photo50 = getString(cursor, PHOTO_50)
            photo100 = getString(cursor, PHOTO_100)
            photo200 = getString(cursor, PHOTO_200)
            isOnline = getInt(cursor, ONLINE) == 1
            isOnlineMobile = getInt(cursor, ONLINE_MOBILE) == 1
            status = getString(cursor, STATUS)
            isVerified = getInt(cursor, VERIFIED) == 1
            lastSeen = getInt(cursor, LAST_SEEN)
            lastSeenPlatform = getInt(cursor, PLATFORM)
        }
    }

    private fun parseGroup(cursor: Cursor): VKGroup {
        return VKGroup().apply {
            id = abs(getInt(cursor, GROUP_ID))
            name = getString(cursor, NAME)
            screenName = getString(cursor, SCREEN_NAME)
            isClosed = getInt(cursor, IS_CLOSED) == 1
            deactivated = getString(cursor, DEACTIVATED)
            photo50 = getString(cursor, PHOTO_50)
            photo100 = getString(cursor, PHOTO_100)
            photo200 = getString(cursor, PHOTO_200)
        }
    }

    private fun putValues(message: VKMessage, values: ContentValues) {
        values.put(MESSAGE_ID, message.id)
        values.put(DATE, message.date)
        values.put(PEER_ID, message.peerId)
        values.put(FROM_ID, message.fromId)
        values.put(EDIT_TIME, message.editTime)
        values.put(OUT, if (message.isOut) 1 else 0)
        values.put(CONVERSATION_MESSAGE_ID, message.conversationMessageId)
        values.put(TEXT, message.text)
        values.put(RANDOM_ID, message.randomId)
        values.put(READ, message.isRead)
        values.put(IMPORTANT, message.isImportant)
        values.put(ATTACHMENTS, Util.serialize(message.attachments))
        values.put(FWD_MESSAGES, Util.serialize(message.fwdMessages))
        values.put(REPLY_MESSAGE_ID, message.replyMessageId)
        values.put(ACTION, Util.serialize(message.action))
    }

    private fun putValues(conversation: VKConversation, values: ContentValues) {
        values.put(PEER_ID, conversation.id)
        values.put(IN_READ, conversation.inRead)
        values.put(OUT_READ, conversation.outRead)
        values.put(UNREAD_COUNT, conversation.unreadCount)
        values.put(TYPE, conversation.type)
        values.put(IS_ALLOWED, conversation.isAllowed)
        values.put(LOCAL_ID, conversation.localId)
        values.put(DISABLED_UNTIL, conversation.disabledUntil)
        values.put(IS_DISABLED_FOREVER, conversation.isDisabledForever)
        values.put(IS_NO_SOUND, conversation.isNoSound)
        values.put(MEMBERS_COUNT, conversation.membersCount)
        values.put(TITLE, conversation.title ?: "")
        values.put(PINNED_MESSAGE_ID, conversation.pinnedMessageId)
        values.put(STATE, conversation.state)
        values.put(IS_GROUP_CHANNEL, conversation.isGroupChannel)
        values.put(PHOTO_50, conversation.photo50)
        values.put(PHOTO_100, conversation.photo100)
        values.put(PHOTO_200, conversation.photo200)
        values.put(LAST_MESSAGE_ID, conversation.lastMessageId)
    }

    private fun putValues(user: VKUser, values: ContentValues, isFriend: Boolean) {
        if (isFriend) {
            values.put(USER_ID, UserConfig.userId)
            values.put(FRIEND_ID, user.id)
            return
        }
        values.put(USER_ID, user.id)
        values.put(FIRST_NAME, user.firstName)
        values.put(LAST_NAME, user.lastName)
        values.put(DEACTIVATED, user.deactivated)
        values.put(CLOSED, user.isClosed)
        values.put(CAN_ACCESS_CLOSED, user.isCanAccessClosed)
        values.put(SEX, user.sex)
        values.put(SCREEN_NAME, user.screenName)
        values.put(PHOTO_50, user.photo50)
        values.put(PHOTO_100, user.photo100)
        values.put(PHOTO_200, user.photo200)
        values.put(ONLINE, user.isOnline)
        values.put(ONLINE_MOBILE, user.isOnlineMobile)
        values.put(STATUS, user.status)
        values.put(LAST_SEEN, user.lastSeen)
        values.put(PLATFORM, user.lastSeenPlatform)
        values.put(VERIFIED, user.isVerified)
    }

    private fun putValues(group: VKGroup, values: ContentValues) {
        values.put(GROUP_ID, abs(group.id))
        values.put(NAME, group.name)
        values.put(SCREEN_NAME, group.screenName)
        values.put(IS_CLOSED, group.isClosed)
        values.put(DEACTIVATED, group.deactivated)
        values.put(TYPE, group.type)
        values.put(PHOTO_50, group.photo50)
        values.put(PHOTO_100, group.photo100)
        values.put(PHOTO_200, group.photo200)
    }

    fun insertConversation(conversation: VKConversation) {
        insertConversations(ArrayList(listOf(conversation)))
    }

    fun insertConversations(conversations: ArrayList<VKConversation>) {
        insert(TABLE_CONVERSATIONS, conversations)
    }

    fun updateConversations(conversations: ArrayList<VKConversation>) {
        val ids = Array(conversations.size) {it.toString()}

        for (i in conversations.indices) {
            ids[i] = conversations[i].id.toString()
        }

        update(TABLE_CONVERSATIONS, conversations, PEER_ID, ids)
    }

    fun deleteConversations(conversations: ArrayList<VKConversation>) {
        val ids = IntArray(conversations.size)

        for (i in conversations.indices) {
            ids[i] = conversations[i].id
        }

        deleteConversations(ids)
    }

    fun deleteConversations(conversationsIds: IntArray) {
        val args = TextUtils.join(", ", conversationsIds.toMutableList())

        delete(TABLE_CONVERSATIONS, PEER_ID, args)
    }

    fun deleteConversation(conversation: VKConversation) {
        deleteConversations(ArrayList(listOf(conversation)))
    }

    fun deleteConversation(conversationId: Int) {
        deleteConversations(intArrayOf(conversationId))
    }

    fun insertUser(user: VKUser) {
        insertUsers(ArrayList(listOf(user)))
    }

    fun insertUsers(users: ArrayList<VKUser>) {
        insert(TABLE_USERS, users)
    }

    fun insertFriends(users: ArrayList<VKUser>) {
        insertUsers(users)
        insert(TABLE_FRIENDS, users)
    }

    fun insertGroup(group: VKGroup) {
        insertGroups(ArrayList(listOf(group)))
    }

    fun insertGroups(groups: ArrayList<VKGroup>) {
        insert(TABLE_GROUPS, groups)
    }

    fun updateMessages(messages: ArrayList<VKMessage>) {
        val ids = Array(messages.size) {it.toString()}

        for (i in messages.indices) {
            ids[i] = messages[i].id.toString()
        }

        update(TABLE_MESSAGES, messages, MESSAGE_ID, ids)
    }

    fun updateMessage(message: VKMessage) {
        updateMessages(ArrayList(listOf(message)))
    }

    fun insertMessages(messages: ArrayList<VKMessage>) {
        insert(TABLE_MESSAGES, messages)
    }

    fun insertMessage(message: VKMessage) {
        insertMessages(ArrayList(listOf(message)))
    }

    fun deleteMessages(messages: ArrayList<VKMessage>) {
        val ids = IntArray(messages.size)

        for (i in messages.indices) {
            ids[i] = messages[i].id
        }

        deleteMessages(ids)
    }

    fun deleteMessages(messagesIds: IntArray) {
        val args = TextUtils.join(", ", messagesIds.toMutableList())

        delete(TABLE_MESSAGES, MESSAGE_ID, args)
    }

    fun deleteMessage(message: VKMessage) {
        deleteMessages(ArrayList(listOf(message)))
    }

    fun deleteMessage(messageId: Int) {
        deleteMessages(intArrayOf(messageId))
    }
}